atuin-client/src/import/zsh_histdb.rs

```diff
@@ -0,0 +1,165 @@
+// import old shell history from zsh-histdb!
+// automatically hoover up all that we can find
+
+// As far as i can tell there are no version numbers in the histdb sqlite DB, so we're going based
+// on the schema from 2022-05-01
+//
+// I have run into some histories that will not import b/c of non UTF-8 characters. 
+//
+
+//
+// An Example sqlite query for hsitdb data: 
+//
+//id|session|command_id|place_id|exit_status|start_time|duration|id|argv|id|host|dir
+//
+//
+//  select 
+//    history.id,
+//    history.start_time,
+//    places.host,
+//    places.dir,
+//    commands.argv 
+//  from history 
+//    left join commands on history.command_id = commands.rowid 
+//    left join places on history.place_id = places.rowid ;
+//
+// CREATE TABLE history  (id integer primary key autoincrement,
+//                       session int,
+//                       command_id int references commands (id),
+//                       place_id int references places (id),
+//                       exit_status int,
+//                       start_time int,
+//                       duration int);
+//
+
+use std::{
+    path::{Path, PathBuf},
+};
+
+use chrono::{prelude::*, Utc};
+use async_trait::async_trait;
+use directories::UserDirs;
+use eyre::{eyre, Result};
+use sqlx::sqlite::SqlitePool;
+
+use super::Importer;
+use crate::history::History;
+use crate::import::Loader;
+
+#[derive(sqlx::FromRow, Debug)]
+pub struct HistDbEntryCount {
+    pub count: usize
+}
+
+#[derive(sqlx::FromRow, Debug)]
+pub struct HistDbEntry {
+    pub id: i64,
+    pub start_time: NaiveDateTime,
+    pub host: String,
+    pub dir: String,
+    pub argv: String,
+    pub duration: i64,
+}
+
+impl From<HistDbEntry> for History {
+    fn from(histdb_item: HistDbEntry) -> Self {
+        History::new (
+            DateTime::from_utc(histdb_item.start_time, Utc), // must assume UTC? 
+            histdb_item.argv.trim_end().to_string(),
+            histdb_item.dir,
+            0, // assume 0, we have no way of knowing :(
+            histdb_item.duration,
+            None,
+            Some(histdb_item.host),
+        )
+    }
+}
+
+#[derive(Debug)]
+pub struct ZshHistDb {
+    histdb: Vec<HistDbEntry>,
+}
+
+
+/// Read db at given file, return vector of entries.
+async fn hist_from_db(dbpath: PathBuf) -> Result<Vec<HistDbEntry>> {
+    let pool = SqlitePool::connect(dbpath.to_str().unwrap()).await?;
+    let query = format!("select history.id,history.start_time,history.duration,places.host,places.dir,commands.argv from history left join commands on history.command_id = commands.rowid left join places on history.place_id = places.rowid order by history.start_time");
```

Comment:
Does this need to be a format? I believe you can pass a string literal to the query_as function

Reply:
clippy has flagged this up too

Reply:
Generally for testing sqlite, we'd open an in-memory instance (open the path ":memory:"), insert some test values, and then run queries against that. That way there's no dependency on the local filesystem

Reply:
Test re-written to eliminate local filesystem dependency. I ended up having to split a few of the functions into small chunks to try to exercise as much code as possible during the tests.